### PR TITLE
Add auth0 provider

### PR DESCRIPTION
**Pull request by Capitrium**
## Description

Adds an Auth0 provider and related docs.

## Motivation and Context

We wanted to add SSO to some of our internal apps, and were already using Auth0.

## How Has This Been Tested?

Included unit tests, and using the Auth0 provider in our Kubernetes cluster.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,9 +1,10 @@
 # Vx.x.x (Pre-release)
 
 ## Release Hightlights
+- Added Auth0 provider
 
 ## Important Notes
-- [#335] The session expiry for the OIDC provider is now taken from the Token Response (expires_in) rather than from the id_token (exp) 
+- [#335] The session expiry for the OIDC provider is now taken from the Token Response (expires_in) rather than from the id_token (exp)
```
Comment (nickveenhof) [inline CHANGELOG.md]: please remove unnecessary whitespaces

File: providers/auth0.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	p.ProviderName = "Auth0"
+	p.Scope = "openid profile email"
+	return &Auth0Provider{ProviderData: p}
+}
+
+func (p *Auth0Provider) Configure(domain string) {
+	if domain == "" {
+		panic(fmt.Sprintf("auth0 domain not set"))
+	}
+	p.Domain = domain
```
Comment (nickveenhof) [inline providers/auth0.go]: Why do you need another param? It simplifies it indeed but it adds another generic option only used for oauth. Looks like simplifying it only makes documentation more complex. Maybe in another PR?

File: options.go [hunk 3]
```diff
@@ -52,6 +52,7 @@ type Options struct {
 	GitHubOrg                string   `flag:"github-org" cfg:"github_org" env:"OAUTH2_PROXY_GITHUB_ORG"`
 	GitHubTeam               string   `flag:"github-team" cfg:"github_team" env:"OAUTH2_PROXY_GITHUB_TEAM"`
 	GitLabGroup              string   `flag:"gitlab-group" cfg:"gitlab_group" env:"OAUTH2_PROXY_GITLAB_GROUP"`
+	Auth0Domain              string   `flag:"auth0-domain" cfg:"auth0_domain" env:"OAUTH2_PROXY_AUTH0_DOMAIN"`
```
Comment (nickveenhof) [inline options.go]: See other comment about this. I think this should not be needed

**Comment by JoelSpeed** [comment]
Is there any benefit to adding this provider over using the existing OIDC provider? Auth0 should support this https://auth0.com/docs/protocols/oidc/openid-connect-discovery

**Comment by starkers** [comment]
I can confirm oidc+auth0 works

**Comment by Capitrium** [comment]
I managed to get Auth0+OIDC working as well, so I'll close this PR; adding a note to the documentation indicating that OIDC is preferred over adding a provider might be a good idea, as there appear to be several existing providers that also support OIDC :wink: ([Google](https://developers.google.com/identity/protocols/OpenIDConnect), [Azure](https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/v1-protocols-openid-connect-code), [Keycloak](https://www.keycloak.org/docs/latest/securing_apps/#openid-connect-2), [GitLab](https://docs.gitlab.com/ee/integration/openid_connect_provider.html))

**Comment by syscll** [comment]
Thanks @Capitrium! Please feel free to contribute again in the future :smile:

**Pull request closed without merging**